rfcs/0026-memory-control.md

```diff
@@ -0,0 +1,113 @@
+
+### Action: How to release the memory?
+
+- **Streaming**: We have implemented a CN-level LRU Manager to control the watermark epoch of all stateful operators. Once memory usage reaches the limit, the watermark epoch will be increased to evict some least-recent entries. See [RFC: Yet another simple idea for memory management](https://singularity-data.quip.com/CldAAcFmzZSO/Yet-another-simple-idea-for-memory-management) for details.
+- **Batch**: We release the memory from batch engine by simply killing 1 or more running queries with the largest memory consumption. The implementation of `kill()` should be identical to handling `Ctrl+C`. I think an `AtomicBool` can make it work.
+
+Notice that the background coroutine works in an async style, that is, there is no strict guarantee that a task must be killed **immediately** once the memory usage exceeds the limit. 
+
+### Policy: Which component should give out memory?
+
+#### Without Overselling
+
+2 straight-forward options are exposed to users:
+
+- `batch_memory_limit_mb`: The **upper bound** of memory consumed by batch executors
+- `streaming_memory_limit_mb`: The **upper bound** of memory consumed by streaming executors
+
+![](images/policy-without-overselling.drawio.svg)
+
+Once the memory usage exceeds the limit, we run the actions to release some memory from streaming or batch.
+
+#### With Overselling (Optional)
+
+In total, there are `batch_memory_limit_mb + streaming_memory_limit_mb` for both streaming and batch executors. @liurenjie1024 proposed to utilize the memory space by allowing streaming and batch executors to use memory from each other, but must return it back immediately once the "owner" requires.
+
+2 additional parameters need to be introduced:
+
+- `batch_memory_reserved_mb`: The memory that must be reserved for batch no matter whether it's free. Must be `<= batch_memory_limit_mb`
+- `streaming_memory_reserved_mb`: The memory that must be reserved for streaming no matter whether it's free. Must be `<= streaming_memory_limit_mb`
+
+![](images/policy-with-overselling.drawio.svg)
+
+Then `free - reserved` is the memory size that could be borrowed by each other, denoted as `overflow`. For example, assuming `limit` of streaming is 500MB and `reserved` is 100MB:
+
+- If `allocated` = 0MB and `free` = 500MB, then `overflow` = 500MB - 100MB = 400MB
+- If `allocated` = 300MB and `free`= 200MB, then `overflow` = 200MB - 100MB = 100MB
+- If `allocated` = 450MB and `free`= 50MB, then `overflow` = 50MB - 100M = -50MB, so overflow is prohibited now
+
+Why `reserved` is necessary? This is because we are using an async way to control memory i.e. with a background coroutine. As a result, **any actions will be slightly later than the actual exhaustion of memory**. The `reserved` memory is designed to mitigate the problem by reserving some space in case that memory cannot be reclaimed immediately.
+
+@fuyufjh prefers the solution without overselling to keep it simple and stupid.
```

Comment:
I think my diagram is clearer for this case. The algorithm is quite intuitive:
```
if (batch_mem_usage + streaming_mem_usage) >= allowed_mem_usage:
    if batch_mem_usage > (allowed_mem_usage - streaming_mem_reserved):
        reject_new_task();
        kill_batch_task();
    else if batch_mem_usage > batch_mem_reserved:
        reject_new_task();
    else:
        evit_streaming_mem();           
```
A simple illustration for dividing memory areas:
| batch_mem_reserved | dynamic_part | streaming_mem_reserved|

With this simple modification, we can be elastic to more cases and utilize memory better. I really don't understand why this is compliecated.

Reply:
Yes, I have re-designed this part and I think the complexity is acceptable to me. But I still prefer the "without overselling" solution, because it's super simple! 🤪